pytorch_lightning/trainer/trainer.py

```diff
@@ -1069,20 +1070,33 @@ def __attach_dataloaders(self, model, train_dataloader=None, val_dataloaders=Non
         if test_dataloaders is not None:
             model.test_dataloader = _PatchDataLoader(test_dataloaders)
 
-    def __attach_datamodule(self, model, datamodule=None):
+    def __attach_datamodule(self, model, datamodule, stage):
 
         # We use datamodule if it's been provided on .fit or .test, otherwise we check model for it
         datamodule = datamodule or getattr(model, 'datamodule', None)
 
         # If we have a datamodule, attach necessary hooks + dataloaders
         if datamodule:
+
+            # If datamodule.prepare_data() has not been called yet, call it
+            if self.is_overridden('prepare_data', datamodule) and not datamodule.has_prepared_data:
+                datamodule.prepare_data()
+
+            # If datamodule.setup('test') has not been called yet, call it
+            if stage == 'test':
+                if self.is_overridden('setup', datamodule) and not datamodule.has_setup_test:
+                    datamodule.setup('test')
+
+            # Override loader hooks
             if self.is_overridden('train_dataloader', datamodule):
                 model.train_dataloader = datamodule.train_dataloader
             if self.is_overridden('val_dataloader', datamodule):
                 model.val_dataloader = datamodule.val_dataloader
             if self.is_overridden('test_dataloader', datamodule):
                 model.test_dataloader = datamodule.test_dataloader
```

Comment:
```suggestion
            if stage == 'fit':
                if self.is_overridden('train_dataloader', datamodule):
                    model.train_dataloader = datamodule.train_dataloader
                if self.is_overridden('val_dataloader', datamodule):
                    model.val_dataloader = datamodule.val_dataloader
            elif stage == 'test':
                if self.is_overridden('test_dataloader', datamodule):
                    model.test_dataloader = datamodule.test_dataloader
```

Reply:
yup. Bonus points if we check for `self.dm`, which lets users avoid passing `trainer.test(datamodule=dm)`